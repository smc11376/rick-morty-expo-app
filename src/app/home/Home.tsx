import { FlatList, StyleSheet, View } from "react-native";
import { useGetAllCharacters } from "@/api/character";
import { useCallback, useState } from "react";
import CharacterCard from "@/components/character-card";
import { ShortCharacter } from "@/domain-models/character";
import { goToCharacterDetail } from "@/libs/navigation";
import HomeHeader from "@/app/home/components/HomeHeader";
import Loader from "@/components/loader";

export default function Home() {
  const [currentPage, setCurrentPage] = useState<string | null | undefined>(null);
  const { characters, isLoading, nextPage } = useGetAllCharacters(currentPage);

  const renderCharacterCard = useCallback(
    (item: ShortCharacter) => (
      <CharacterCard character={item} onPressCharacter={() => goToCharacterDetail(item.id, item.name)} />
    ),
    [characters],
  );

  function handleLoadMore() {
    !!nextPage && setCurrentPage(nextPage);
  }

  return (
    <View style={styles.container}>
      <HomeHeader />
      <FlatList
        data={characters}
        renderItem={({ item }) => renderCharacterCard(item)}
        keyExtractor={item => item.id.toString()}
        onEndReached={handleLoadMore}
        onEndReachedThreshold={0.3}
        ListFooterComponent={() => (isLoading ? <Loader /> : null)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "rgb(39, 43, 51)",
  },
});
