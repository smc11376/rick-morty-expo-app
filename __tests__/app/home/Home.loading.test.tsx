import { render, screen } from "@testing-library/react-native";
import Home from "@/app/home/Home";

jest.mock("@/api/character", () => ({
  useGetAllCharacters: () => ({
    characters: [],
    isLoading: false,
    nextPage: null,
  }),
}));

test("Home renders isLoading", () => {
  render(<Home />);

  expect(screen.toJSON()).toMatchSnapshot();
});
