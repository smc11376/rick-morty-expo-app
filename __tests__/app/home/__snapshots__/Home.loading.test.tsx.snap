// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders isLoading 1`] = `
<View
  style={
    {
      "backgroundColor": "rgb(39, 43, 51)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "height": 160,
          "justifyContent": "center",
          "width": "100%",
        },
        {
          "paddingTop": 20,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      The Rick and Morty API
    </Text>
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.3}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</View>
`;
